The CI errors are strange, and it's not clear right away if they're easily fixed:

```
  Error: Coverage upload failed: Coverage upload failed (HTTP 403): Code quality is not enabled for this repository. Enable code quality in your repository settings.. To treat upload errors as warnings, add 'fail-on-error: false' to the action inputs. See https://docs.github.com/en/code-security/how-tos/maintain-quality-code/set-up-code-coverage for more information.
  Error: Process completed with exit code 1.
```

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,4 +1,12 @@
-version: "2"
+version: 2
+
+run:
+  timeout: 10m
+
+linters-settings:
+  golint:
+    min-confidence: 0.9
+
 linters:
   default: none
   enable:

--- a/channel/combined.go
+++ b/channel/combined.go
@@ -52,6 +52,19 @@ func (c *CombinedSource) Update(ctx context.Context, logger *logrus.Entry) error
 	return nil
 }
 
+func (c *CombinedSource) GarbageCollect(ctx context.Context, logger *logrus.Entry) error {
+	for _, source := range c.sources {
+		if gc, ok := source.(GarbageCollector); ok {
+			err := gc.GarbageCollect(ctx, logger)
+			if err != nil {
+				return fmt.Errorf("error while garbage collecting source %s: %v", source.Name(), err)
+			}
+		}
+	}
+
+	return nil
+}
+
 func (c *CombinedSource) Version(channel string, overrides map[string]string) (ConfigVersion, error) {
 	versions := make([]ConfigVersion, len(c.sources))
 	for i, source := range c.sources {

--- a/channel/config_source.go
+++ b/channel/config_source.go
@@ -28,6 +28,13 @@ type ConfigSource interface {
 	Version(channel string, overrides map[string]string) (ConfigVersion, error)
 }
 
+// GarbageCollector is an optional interface for config sources that support
+// garbage collection. Sources should implement this to support deferred
+// garbage collection after all updates are complete.
+type GarbageCollector interface {
+	GarbageCollect(ctx context.Context, logger *log.Entry) error
+}
+
 type Manifest struct {
 	Path     string
 	Contents []byte

--- a/channel/git.go
+++ b/channel/git.go
@@ -127,9 +127,28 @@ func (g *Git) Update(ctx context.Context, logger *log.Entry) error {
 	return nil
 }
 
+// GarbageCollect performs garbage collection on the git repository.
+// This should be called after all Update operations have completed to avoid
+// blocking on individual update calls.
+func (g *Git) GarbageCollect(ctx context.Context, logger *log.Entry) error {
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+
+	_, err := os.Stat(g.repoDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	return g.cmd(ctx, logger, "--git-dir", g.repoDir, "gc", "--aggressive")
+}
+
 func (g *Git) Version(channel string, _ map[string]string) (ConfigVersion, error) {
 	stderr := new(bytes.Buffer)
-	cmd := exec.Command("git", "--git-dir", g.repoDir, "rev-parse", channel)
+	gitArgs := []string{"-c", "gc.auto=0", "--git-dir", g.repoDir, "rev-parse", channel}
+	cmd := exec.Command("git", gitArgs...)
 	cmd.Stderr = stderr
 	sha, err := cmd.Output()
 	if err != nil {
@@ -167,8 +186,13 @@ func (g *Git) localClone(ctx context.Context, logger *log.Entry, channel string)
 }
 
 // cmd executes a git command with the correct environment set.
+// Disables automatic garbage collection to prevent blocking during concurrent operations.
 func (g *Git) cmd(ctx context.Context, logger *log.Entry, args ...string) error {
-	cmd := exec.Command("git", args...)
+	gitArgs := make([]string, 0, len(args)+2)
+	gitArgs = append(gitArgs, "-c", "gc.auto=0")
+	gitArgs = append(gitArgs, args...)
+
+	cmd := exec.Command("git", gitArgs...)
 	// set GIT_SSH_COMMAND with private-key file when pulling over ssh.
 	if g.sshPrivateKeyFile != "" {
 		cmd.Env = []string{fmt.Sprintf("GIT_SSH_COMMAND=ssh -i %s -o 'StrictHostKeyChecking no'", g.sshPrivateKeyFile)}

--- a/channel/git_test.go
+++ b/channel/git_test.go
@@ -98,6 +98,30 @@ func TestGitGet(t *testing.T) {
 	verifyExampleConfig(t, sha, "testsrc", "channel1")
 }
 
+func TestGitGarbageCollect(t *testing.T) {
+	logger := log.WithFields(map[string]interface{}{})
+
+	repoTempDir := CreateTempDir(t)
+	defer os.RemoveAll(repoTempDir)
+
+	workDir := CreateTempDir(t)
+	defer os.RemoveAll(workDir)
+
+	createGitRepo(t, logger, repoTempDir)
+	c, err := NewGit(command.NewExecManager(1), "testsrc", workDir, repoTempDir, "")
+	require.NoError(t, err)
+
+	err = c.Update(context.Background(), logger)
+	require.NoError(t, err)
+
+	// GarbageCollect should succeed without errors
+	gc, ok := c.(GarbageCollector)
+	require.True(t, ok)
+
+	err = gc.GarbageCollect(context.Background(), logger)
+	require.NoError(t, err)
+}
+
 func TestGetRepoName(t *testing.T) {
 	for _, tc := range []struct {
 		msg     string

--- a/controller/controller.go
+++ b/controller/controller.go
@@ -126,6 +126,13 @@ func (c *Controller) refresh() error {
 		return err
 	}
 
+	if gc, ok := c.channelConfigSourcer.(channel.GarbageCollector); ok {
+		err = gc.GarbageCollect(context.Background(), c.logger)
+		if err != nil {
+			c.logger.Warnf("failed to perform garbage collection: %v", err)
+		}
+	}
+
 	clusters, err := c.registry.ListClusters(
 		registry.Filter{
 			Providers: c.providers,

